### PR TITLE
Add `kokkos_swap(Array<T, N>)` specialization

### PR DESCRIPTION
**Pull request by dalg24**
Specializing the swap algorithm for Kokkos arrays was initially proposed in #6697 but we dropped it to focus on the Kokkos swap ADL ordeal. Somehow we overlooked a stray <Kokkos_Swap.hpp> header include in the Kokkos::Array header file.  This PR reintroduce a
`Kokkos::kokkos_swap(Kokkos::Array)` specialization, following closely what the standard library does for `std::swap(std::array)`.

### CODE REVIEW

**Review by Rombur — APPROVED** [review]

File: core/src/Kokkos_Array.hpp [hunk 1]
```diff
@@ -185,6 +196,10 @@ struct Array<T, 0> {
   // for default move constructor and move assignment operator.
   // Array( Array && ) = default ;
   // Array & operator = ( Array && ) = default ;
+
+ private:
+  friend KOKKOS_INLINE_FUNCTION constexpr void kokkos_swap(
+      Array<T, 0>&, Array<T, 0>&) noexcept(Impl::is_nothrow_swappable_v<T>) {}
```
Comment (nliber) [inline core/src/Kokkos_Array.hpp:202]: Why is this dependent on `is_nothrow_swappable<T>` and not just `noexcept(true)`?

It should have the same exception specification as [[array.zero]](https://eel.is/c++draft/array.zero#4).

**Review by nliber — DISMISSED** [review]

File: core/src/Kokkos_Array.hpp [hunk 2]
```diff
@@ -133,6 +133,17 @@ struct Array {
   KOKKOS_INLINE_FUNCTION constexpr const_pointer data() const {
     return &m_internal_implementation_private_member_data[0];
   }
+
+ private:
+  template <class U = T>
```
Comment (maartenarnst) [inline core/src/Kokkos_Array.hpp:138]: It seems the template argument `U` is not used (?). I may be missing something.

Comment (dalg24) [reply core/src/Kokkos_Array.hpp:138]: You're right I should have used in the SFINAE condition below instead of `T`

Comment (dalg24) [reply core/src/Kokkos_Array.hpp:138]: Not sure why the compiler didn't yell at me

File: core/src/Kokkos_Array.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/Kokkos_Array.hpp:202]: You're right, I had looked in cppreference and read the framed `noexcept` specifications and missed the next clause that clearly stated:

> For zero-sized arrays, [noexcept](https://en.cppreference.com/w/cpp/language/noexcept_spec) specification:  noexcept

Will fix.  Thanks.

**Review by crtrott — APPROVED** [review]

**Pull request merged**